### PR TITLE
secp256k1/ecdsa: Add sign and verify tests.

### PR DESCRIPTION
**Pull request by davecgh**
Although the ECDSA code has been extensively tested through usage and external fuzz testing, there are not currently any unit tests for the non-compact signing and verification path.

This resolves that by adding signing and verification tests for both the positive and negative paths.

The following is a high-level overview of the changes:

- Refactors the primary logic for the regular ECDSA signing code to a separate function that accepts an arbitrary nonce to facilitate fully testing both the positive and negative paths and provides more flexibility for potentially using other nonce generation mechanisms in the future should it be necessary
- Add sign and verify tests with known good test data that is easily reproducible and has been independently verified with the Sage computer algebra system
  - The test data includes the original messages so the resulting hashes are independently verifiable and have a known source
  - The test data includes variations of signing the same data with different keys and nonces, both deterministically generated via RFC6979 and random, and signing different data with the same keys
- Add tests to ensure the aforementioned test data that was independently verified produces the expected results
- Add tests which are specifically designed to exercise conditions that lead to invalid signatures
- Add tests to help ensure the signature verification code fails to verify edge conditions as expected by crafting signatures that are specifically designed to hit them
- Add test to sign and verify random data
    - Generates random keys and messages with each run from a new random seed and log that seed in the event of failure
    - Signs the rand message with the random key and ensures the produced signature verifies correctly
    - Ensures mutating a random bit in each good signature results in that mutated signature failing to verify the original message
    - Ensures mutating a random bit in each message hash that was originally signed results in the original good signature failing to verify the new mutated message

### CODE REVIEW

File: dcrec/secp256k1/ecdsa/signature.go [hunk 1]
```diff
@@ -581,82 +703,14 @@ func signRFC6979(privKey *secp256k1.PrivateKey, hash []byte) (*Signature, byte)
...
-		// Repeat from step 1 if s = 0
-		// s = -s if s > N/2
-		kInv := new(secp256k1.ModNScalar).InverseValNonConst(k)
+		// Steps 2-6.
+		sig, pubKeyRecoveryCode, success := sign(privKeyScalar, k, hash)
 		k.Zero()
-		s := new(secp256k1.ModNScalar).Mul2(privKeyScalar, &r).Add(&e).Mul(kInv)
-		if s.IsZero() {
+		if !success {
 			continue
```
Comment (matthawkins90) [inline dcrec/secp256k1/ecdsa/signature.go:710]: If the sign() function generates an `r` or `s` with a value of zero, that function yields "success = false".
According to the algorithm notes, you must start over from step 1 (a new nonce) in this case.
By continuing on to the next iteration of this for loop, the iteration is incremented, a new nonce is calculated, and it tries the sign() function again.

Just want to make sure I understand the iteration for loop correctly. Seems like each individual iteration isn't important, and this will use as many iterations as required (though probably only 1 attempt) to generate a valid sig & recovery code. Correct?

This could all be apparent already, but a comment here might make it a bit more clear:
> // Retry from step 1 (using a new nonce)

Comment (davecgh) [reply dcrec/secp256k1/ecdsa/signature.go:710]: Yes, that's correct.  Maybe it's not quite as obvious from the diff due to being interspersed with the moved code, but it seems pretty apparent from the final code to me since it's only 4 lines of code away from the iteration loop and the comment on the nonce call says `Generate a deterministic nonce in [1, N-1] parameterized by the private key, message being signed, and iteration count.`

I can add it if you really feel it's non-obvious, but I suspect it would be redundant in the final code due to its proximity.

Comment (matthawkins90) [reply dcrec/secp256k1/ecdsa/signature.go:710]: Nope, I initially didn't realize that the entire point of the for loop is to keep trying nonces until one works. Makes more sense now. I think it just took me a while to understand, and will definitely be more clear outside of the diff.

**Review by JoeGruffins — APPROVED** [review]

File: dcrec/secp256k1/ecdsa/signature_test.go [hunk 2]
```diff
@@ -271,6 +274,464 @@ func TestSignatureSerialize(t *testing.T) {
...
+			t.Errorf("%s: unexpected success -- got sig %x", test.name,
+				sig.Serialize())
+			continue
+		}
+	}
+}
+
+// TestVerifyFailures ensures the ECDSA verification function returns an
+// unsuccessful result for edge conditions.
+func TestVerifyFailures(t *testing.T) {
```
Comment (rstaudt2) [inline dcrec/secp256k1/ecdsa/signature_test.go:686]: It doesn't make a big difference since there are not a huge amount of tests, but it might be worth running the tests in parallel as a standard since it will eventually add up across all the various test suites:
```suggestion
func TestVerifyFailures(t *testing.T) {
    t.Parallel()
```

**Review by rstaudt2 — APPROVED** [review]
Nice increase in coverage!

File: dcrec/secp256k1/ecdsa/signature_test.go [hunk 3]
```diff
@@ -271,6 +274,464 @@ func TestSignatureSerialize(t *testing.T) {
...
+		rfc6979:  true,
+		wantSigR: "0d4cbf2da84f7448b083fce9b9c4e1834b5e2e98defcec7ec87e87c739f5fe78",
+		wantSigS: "0997db60683e12b4494702347fc7ae7f599e5a95c629c146e0fc615a1a2acac5",
+		wantCode: pubKeyRecoveryCodeOddnessBit,
+	}}
+
+	// Ensure the test data is sane by comparing the provided hashed message and
+	// nonce, in the case RFC6979 was used, to their calculated values.  These
+	// values could just be calculated instead of specified in the test data,
+	// but it's nice to have all of the calcuated values available in the test
```
[the remark this answers is not in the document]
Comment (davecgh) [reply dcrec/secp256k1/ecdsa/signature_test.go]: Corrected.

File: dcrec/secp256k1/ecdsa/signature_test.go (hunk 2 above)
Comment (davecgh) [reply dcrec/secp256k1/ecdsa/signature_test.go:692]: Agreed.  I added it to all of the new tests introduced as well.
